clients/git/client.go

```diff
@@ -0,0 +1,208 @@
+			URL:      uri,
+			Progress: os.Stdout,
+		})
+		if err != nil {
+			return fmt.Errorf("git.PlainClone: %w", err)
+		}
+	}
+	c.tempDir = tempDir
+	c.worktree, err = c.gitRepo.Worktree()
+	if err != nil {
+		return fmt.Errorf("git.Worktree: %w", err)
+	}
+
+	// git checkout
+	if commitSHA != clients.HeadSHA {
+		if err := c.worktree.Checkout(&git.CheckoutOptions{
+			Hash:  plumbing.NewHash(commitSHA),
+			Force: true, // throw away any unsaved changes.
+		}); err != nil {
+			return fmt.Errorf("git.Worktree: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) ListCommits() ([]clients.Commit, error) {
+	c.listCommits.Do(func() {
+		commitIter, err := c.gitRepo.Log(&git.LogOptions{
+			Order: git.LogOrderCommitterTime,
+		})
+		if err != nil {
+			c.errListCommits = fmt.Errorf("git.CommitObjects: %w", err)
+			return
+		}
+		for i := 0; i < c.commitDepth; i++ {
+			commit, err := commitIter.Next()
+			if err != nil && !errors.Is(err, io.EOF) {
+				c.errListCommits = fmt.Errorf("commitIter.Next: %w", err)
+				return
+			}
+			// No more commits.
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			if commit == nil {
+				// Not sure in what case a nil commit is returned. Fail explicitly.
+				c.errListCommits = fmt.Errorf("%w", errNilCommitFound)
+				return
+			}
+
+			c.commits = append(c.commits, clients.Commit{
+				SHA:           commit.Hash.String(),
+				Message:       commit.Message,
+				CommittedDate: commit.Committer.When,
+				Committer: clients.User{
+					Login: commit.Committer.Email,
+				},
+			})
+		}
+	})
+	return c.commits, c.errListCommits
+}
+
+func (c *Client) Search(request clients.SearchRequest) (clients.SearchResponse, error) {
+	// Pattern
+	if request.Query == "" {
+		return clients.SearchResponse{}, fmt.Errorf("%w", errEmptyQuery)
+	}
+	queryRegexp, err := regexp.Compile(request.Query)
+	if err != nil {
+		return clients.SearchResponse{}, fmt.Errorf("regexp.Compile: %w", err)
+	}
+	grepOpts := &git.GrepOptions{
+		Patterns: []*regexp.Regexp{queryRegexp},
+	}
+
+	// path/filename
+	var pathExpr string
+	switch {
+	case request.Path != "" && request.Filename != "":
+		pathExpr = filepath.Join(fmt.Sprintf("^%s", request.Path),
+			fmt.Sprintf(".*%s", request.Filename))
+	case request.Path != "":
+		pathExpr = fmt.Sprintf("^%s", request.Path)
+	case request.Filename != "":
+		pathExpr = filepath.Join(".*", request.Filename)
+	}
+	if pathExpr != "" {
+		pathRegexp, err := regexp.Compile(pathExpr)
+		if err != nil {
+			return clients.SearchResponse{}, fmt.Errorf("regexp.Compile: %w", err)
+		}
+		grepOpts.PathSpecs = append(grepOpts.PathSpecs, pathRegexp)
+	}
+
+	// Grep
+	grepResults, err := c.worktree.Grep(grepOpts)
+	if err != nil {
+		return clients.SearchResponse{}, fmt.Errorf("git.Grep: %w", err)
+	}
+
+	ret := clients.SearchResponse{}
+	for _, grepResult := range grepResults {
+		ret.Results = append(ret.Results, clients.SearchResult{
```

Comment:
I would be surprised if it is across trees since the API is `Worktree.Grep()`. Also, I did some basic testing and it seems to return results from the single commit SHA that was checked out.